File: code/Java/src/com/longluo/FFT_Recursion_v2.java
package com.longluo;

import java.util.Arrays;
import java.util.NoSuchElementException;
import java.util.Scanner;

/**
 * FFT Recursion Version V2.0
 */
public class FFT_Recursion_v2 {

    static class Solution {
        static class Complex {
            double re;
            double im;

            Complex(double re, double im) {
                this.re = re;
                this.im = im;
            }

            public Complex plus(Complex other) {
                return new Complex(this.re + other.re, this.im + other.im);
            }

            public Complex minus(Complex other) {
                return new Complex(this.re - other.re, this.im - other.im);
            }

            public Complex times(Complex other) {
                double real = this.re * other.re - this.im * other.im;
                double imag = this.re * other.im + this.im * other.re;
                return new Complex(real, imag);
            }

            public Complex scale(double alpha) {
                return new Complex(alpha * re, alpha * im);
            }

            public Complex conjugate() {
                return new Complex(re, -im);
            }

            public String toString() {
                return "(" + re + "," + im + ")";
            }
        }

        public static Complex[] FFT(Complex[] x) {
            int n = x.length;

            // base case
            if (n == 1) {
                return new Complex[]{x[0]};
            }

            if (n % 2 != 0) {
                throw new IllegalArgumentException("n is not a power of 2");
            }

            // fft of even terms
            Complex[] even = new Complex[n / 2];
            for (int k = 0; k < n / 2; k++) {
                even[k] = x[2 * k];
            }
            Complex[] q = FFT(even);

            // fft of odd terms
            Complex[] odd = even;  // reuse the array
            for (int k = 0; k < n / 2; k++) {
                odd[k] = x[2 * k + 1];
            }
            Complex[] r = FFT(odd);

            // combine
            Complex[] y = new Complex[n];
            for (int k = 0; k < n / 2; k++) {
                double kth = -2 * k * Math.PI / n;
                Complex wk = new Complex(Math.cos(kth), Math.sin(kth));
                y[k] = q[k].plus(wk.times(r[k]));
                y[k + n / 2] = q[k].minus(wk.times(r[k]));
            }

            return y;
        }

        public static Complex[] IFFT(Complex[] x) {
            int n = x.length;
            Complex[] y = new Complex[n];

            // take conjugate
            for (int i = 0; i < n; i++) {
                y[i] = x[i].conjugate();
            }

            // compute forward FFT
            y = FFT(y);

            // take conjugate again
            for (int i = 0; i < n; i++) {
                y[i] = y[i].conjugate();
            }

            // divide by n
            for (int i = 0; i < n; i++) {
                y[i] = y[i].scale(1.0 / n);
            }

            return y;
        }

        public static String multiply(String num1, String num2) {
            if ((num1.length() == 1 && Integer.parseInt(num1) == 0)
                    || (num2.length() == 1 && Integer.parseInt(num2) == 0)) {
                return "0";
            }

            int len1 = num1.length();
            int len2 = num2.length();

            int n = 1;
            while (n < len1 + len2) {
                n = n << 1;
            }

            Complex[] A = new Complex[n];
            Complex[] B = new Complex[n];

            Arrays.setAll(A, e -> new Complex(0, 0));
            Arrays.setAll(B, e -> new Complex(0, 0));

            for (int i = len1 - 1; i >= 0; i--) {
                A[len1 - 1 - i].re = num1.charAt(i) - '0';
            }
            for (int i = len2 - 1; i >= 0; i--) {
                B[len2 - 1 - i].re = num2.charAt(i) - '0';
            }

            Complex[] mid1 = FFT(A);
            Complex[] mid2 = FFT(B);

            Complex[] mid = new Complex[n];
            for (int i = 0; i < n; i++) {
                mid[i] = mid1[i].times(mid2[i]);
            }

            Complex[] res = IFFT(mid);

            int[] sum = new int[n];
            int carry = 0;
            for (int i = 0; i < n; i++) {
                int op = (int) (Math.floor(res[i].re + 0.5)) + carry;
                carry = op / 10;
                sum[i] = op % 10;
            }

            int idx = n - 1;
            while (sum[idx] == 0 && idx > 0) {
                idx--;
            }

            StringBuilder ans = new StringBuilder();
            for (int i = idx; i >= 0; i--) {
                ans.append(sum[i]);
            }

            return ans.toString();
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        try {
            while (true) {
                System.out.println("Please Input 2 numbers: ");
                String[] numbers = sc.nextLine().split("\\s+");
                if (numbers.length > 2) {
                    System.out.println("Error Input!");
                    continue;
                }

                Solution s = new Solution();
                String result = s.multiply(numbers[0], numbers[1]);
                System.out.println("Result: " + result + " = " + numbers[0] + " * " + numbers[1]);
            }
        } catch (IllegalStateException | NoSuchElementException e) {
            e.printStackTrace();
        }
    }
}

File: code/cpp/fft_recursion_v2.cpp
//
// Created by longluo on 2022/5/17.
//

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    const double PI = acos(-1.0);  // PI = arccos(-1)

    void display(vector<complex<double>> num) {
        for (auto &cd : num) {
            cout << cd;
        }

        cout << endl;
    }

    /**
     * FFT Recursion 实现
     *
     * @param a
     * @param invert true means IFFT, else FFT
     * @return y
     */
    vector<complex<double>> FFT(vector<complex<double>> &a, bool invert) {
        //第一个参数为一个多项式的系数, 以次数从小到大的顺序, 向量中每一项的实部为该项系数
        int n = a.size();

        // 如果当前多项式仅有常数项时直接返回多项式的值
        if (n == 1) {
            return a;
        }

        vector<complex<double>> Pe(n / 2), Po(n / 2); // 文中的Pe与Po的系数表示法

        for (int i = 0; 2 * i < n; i++) {
            Pe[i] = a[2 * i];
            Po[i] = a[2 * i + 1];
        }

        // Divide 分治
        // 递归求 ye = Pe(xi), yo = Po(xi)
        vector<complex<double>> ye = FFT(Pe, invert);
        vector<complex<double>> yo = FFT(Po, invert);

        // Combine
        vector<complex<double>> y(n);

        // Root of Units
        double ang = 2 * PI / n * (invert ? -1 : 1);
        complex<double> wn(cos(ang), sin(ang)); // wn为第1个n次复根,
        complex<double> w(1, 0);   // w为第零0个n次复根, 即为 1

        for (int i = 0; i < n / 2; i++) {
            y[i] = ye[i] + w * yo[i];  // 求出P(xi)
            y[i + n / 2] = ye[i] - w * yo[i]; // 由单位复根的性质可知第k个根与第k + n/2个根互为相反数
            w *= wn;  // w * wn 得到下一个复根
        }

        // IFFT最终结果需要除以 n ，其中 n 为不小于原多项式阶+1的最小2的整数次幂
//    if (invert) {
//        for (auto &cd : y) {
//            cd /= n;
//        }
//    }

        return y;  // 返回最终的系数
    }

    string multiply(string num1, string num2) {
        if (num1 == "0" || num2 == "0") {
            return "0";
        }

        int len1 = num1.size();
        int len2 = num2.size();

        int n = 1;
        while (n < len1 + len2) {
            n = n << 1;
        }

        vector<complex<double>> a(n);
        vector<complex<double>> b(n);

        for (int i = len1 - 1; i >= 0; i--) {
            a[i] = complex<double>(num1[len1 - 1 - i] - '0', 0);
        }

        for (int i = len2 - 1; i >= 0; i--) {
            b[i] = complex<double>(num2[len2 - 1 - i] - '0', 0);
        }

        cout << "Before FFT" << endl;
        display(a);
        display(b);

        vector<complex<double>> midA, midB;

        midA = FFT(a, false);
        midB = FFT(b, false);

        cout << "After FFT" << endl;
        display(midA);
        display(midB);

        vector<complex<double>> mid(n);
        for (int i = 0; i < n; i++) {
            mid[i] = midA[i] * midB[i];
        }

        cout << "Times:" << endl;
        display(mid);

        vector<complex<double>> result;
        result = FFT(mid, true);

        cout << "Result:" << endl;
        display(result);

        string ans;
        int carry = 0;
        for (int i = 0; i < result.size(); i++) {
            int sum = round(round(result[i].real()) / result.size()) + carry;
            carry = sum / 10;
            ans += sum % 10 + '0';
        }

        if (carry > 0) {
            ans += carry % 10 + '0';
        }

        int idx = ans.size() - 1;
        while (ans[idx] == '0' && idx > 0) {
            idx--;
        }

        ans = ans.substr(0, idx + 1);
        cout << "Output: " << ans << endl;
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

int main() {
    cout << "Please Input 2 Numbers:" << endl;
    string num1, num2;
    cin >> num1 >> num2;

    Solution s;
    string result = s.multiply(num1, num2);

    cout << "Result is:" << endl;
    cout << num1 << " x " << num2 << "=" << result << endl;
}
